File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys

n = 0

visited = []
dx = [0, 0, -1, 1, 1, 1, -1, -1]
dy = [-1, 1, 0, 0, -1, 1, 1, -1]

r = []
a = []
b = []

def main():
    global n
    global cnt
    cnt = 0
    n = int(input())
    for i in range(n):
        r.append(False)
    for i in range(2*n-1):
        a.append(False)
        b.append(False)
    dfs(0)
    print(cnt)

def dfs(idx):
    global cnt
    if idx == n:
        cnt+=1
        return
    for i in range(n):
        if r[i] == True or a[idx + i] == True or b[idx-i] == True:
            continue
        r[i] = True
        a[idx+i] = True
        b[idx-i] = True
        dfs(idx+1)
        r[i] = False
        a[idx+i] = False
        b[idx-i] = False

if __name__ == "__main__":
    main()

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.util.*;
public class Main {
    static boolean[][] visited; // 퀸의 위치
    static int N;
    static int cnt;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        visited = new boolean[N][N];
        dfs(0);
        System.out.println(cnt);
    }

    public static void dfs(int row) {
        if (row == N) {
            cnt++;
            return;
        }
        for (int col = 0 ; col < N; col++){
                if (!isSafe(row, col)) continue;
                visited[row][col] = true;
                dfs(row+1);
                visited[row][col] = false;
        }
    }
    public static boolean isSafe(int row, int col){
        for (int i = 0; i < row; i++){
            if (visited[i][col]) return false;
        }
        // \ 대각선
        int r = row - 1;
        int c = col - 1;
        while (r >= 0 && c >= 0){
            if (visited[r][c]) return false;
            r--;
            c--;
        }

        r = row - 1;
        c = col + 1;
        while (r >= 0 && c < N){
            if (visited[r][c]) return false;
            r--;
            c++;
        }
        return true;
    }
}
